homeassistant/components/sensorpush_cloud/api.py

```diff
@@ -0,0 +1,155 @@
+                # requests are retried before raising an error.
+                if retries < REQUEST_RETRIES:
+                    retries = retries + 1
+                    continue
+
+                LOGGER.debug(f"API call to {func} failed after {retries} retries")
+                if isinstance(e, ApiException):
+                    # API exceptions provide a JSON-encoded message in the
+                    # body; otherwise, fall back to the general behavior.
+                    try:
+                        data = json.loads(e.body)
+                        raise SensorPushCloudError(data["message"]) from e
+                    except Exception:  # noqa: BLE001
+                        pass
+                raise SensorPushCloudError from e
+            else:
+                LOGGER.debug(f"API call to {func} succeeded after {retries} retries")
+                return result
+
+    return _api_call
+
+
+class SensorPushCloudError(Exception):
+    """An exception occurred when calling the SensorPush Cloud API."""
+
+
+class SensorPushCloudApi:
+    """SensorPush Cloud API class."""
+
+    hass: HomeAssistant
+    email: str
+    password: str
+    configuration: Configuration
+    api: ApiApi
+    deadline: datetime
+    lock: Lock
+
+    def __init__(self, hass: HomeAssistant, email: str, password: str) -> None:
+        """Initialize the SensorPush Cloud API object."""
+        self.hass = hass
+        self.email = email
+        self.password = password
+        # Generated Swagger clients install default logging handlers that
+        # conflict with Home Assistant; remove before making API calls.
+        self.configuration = Configuration()
+        for logger in self.configuration.logger.values():
+            logger.removeHandler(self.configuration.logger_stream_handler)
+        self.api = ApiApi(ApiClient(self.configuration))
+        self.deadline = dt_util.now()
+        self.lock = Lock()
+
+    @api_call
+    def authorize(self) -> None:
+        """Sign in and request an authorization code."""
+        # SensorPush provides a simplified OAuth endpoint using access tokens
+        # without refresh tokens. It is not possible to use 3rd party client
+        # IDs without first contacting SensorPush support.
+        response = self.api.oauth_authorize_post(
+            AuthorizeRequest(email=self.email, password=self.password),
+            _request_timeout=REQUEST_TIMEOUT.total_seconds(),
+        )
+        response = self.api.access_token(
+            AccessTokenRequest(authorization=response.authorization),
+            _request_timeout=REQUEST_TIMEOUT.total_seconds(),
+        )
+        self.configuration.api_key["Authorization"] = response.accesstoken
+        self.deadline = dt_util.now() + ACCESS_TOKEN_EXPIRATION
+
+    async def async_authorize(self) -> None:
+        """Sign in and request an authorization code."""
+        return await self.hass.async_add_executor_job(self.authorize)
```

Comment:
How can we make everything async from the getgo?

Reply:
This is a tough one. My preference is for everything to be async as well, however generated Swagger clients have a broad set of compatibility constraints with little support for async. My hope was to at least quarantine blocking calls and then rely on the executor to at least give the appearance of supporting async calls.